File: hub/agents/npm/agent-email/src/platform.ts
// Copyright(C) 2025-2026 Advanced Micro Devices, Inc. All rights reserved.
// SPDX-License-Identifier: MIT
/**
 * Platform/arch resolution and binary-lock loading.
 *
 * The lock file (`binaries.lock.json`, shipped in the published package) is the
 * single source of truth for which artifact to download for the current host
 * and what its SHA-256 must be. Platform keys are `${process.platform}-${process.arch}`.
 */

import { fileURLToPath } from "node:url";
import path from "node:path";
import fs from "node:fs";

import { PlatformError } from "./errors.js";

/** One platform's artifact entry in the lock file. */
export interface BinaryLockEntry {
  /** Artifact filename as published under the base URL, e.g. "email-agent-win32-x64.exe". */
  filename: string;
  /** Lowercase hex SHA-256 of the downloaded artifact. */
  sha256: string;
  /** Size in bytes (informational; not enforced). */
  size?: number;
  /** Basename the executable should be written as on disk (with platform ext). */
  executable: string;
}

/** The whole lock file (`binaries.lock.json`). */
export interface BinaryLock {
  schemaVersion: string;
  agentVersion: string;
  /**
   * Default download base URL. Overridable at fetch time. May be a placeholder
   * if no binaries are published for this build.
   */
  baseUrl: string;
  binaries: Record<string, BinaryLockEntry>;
}

/** Supported platform-arch keys. */
export const SUPPORTED_PLATFORMS = [
  "win32-x64",
  "darwin-arm64",
  "darwin-x64",
  "linux-x64",
] as const;

/** Resolve the current host's platform key, e.g. "win32-x64". */
export function currentPlatformKey(
  platform: NodeJS.Platform = process.platform,
  arch: string = process.arch,
): string {
  return `${platform}-${arch}`;
}

/** Locate `binaries.lock.json` (package root, one level up from dist/ or src/). */
export function defaultLockPath(): string {
  const here = path.dirname(fileURLToPath(import.meta.url));
  // dist/platform.js -> package root is one up; in ts-source (vitest) it's also one up from src/.
  return path.resolve(here, "..", "binaries.lock.json");
}

/** Load and minimally validate the lock file. */
export function loadLock(lockPath: string = defaultLockPath()): BinaryLock {
  let raw: string;
  try {
    raw = fs.readFileSync(lockPath, "utf8");
  } catch (e) {
    throw new PlatformError(
      `cannot read binaries.lock.json at ${lockPath}: ${(e as Error).message}. ` +
        "This manifest ships with the package; reinstall @amd-gaia/agent-email if it is missing.",
    );
  }
  let parsed: BinaryLock;
  try {
    parsed = JSON.parse(raw) as BinaryLock;
  } catch (e) {
    throw new PlatformError(
      `binaries.lock.json at ${lockPath} is not valid JSON: ${(e as Error).message}`,
    );
  }
  if (!parsed.binaries || typeof parsed.binaries !== "object") {
    throw new PlatformError(
      `binaries.lock.json at ${lockPath} is missing a "binaries" map`,
    );
  }
  return parsed;
}

/** Resolve the lock entry for a platform key, failing loudly if unsupported. */
export function resolveEntry(lock: BinaryLock, platformKey: string): BinaryLockEntry {
  const entry = lock.binaries[platformKey];
  if (!entry) {
    const available = Object.keys(lock.binaries).join(", ") || "(none)";
    throw new PlatformError(
      `no email-agent binary for platform '${platformKey}'. ` +
        `Available in binaries.lock.json: ${available}. ` +
        "Supported targets: " +
        SUPPORTED_PLATFORMS.join(", "),
    );
  }
  if (!entry.sha256 || !entry.filename || !entry.executable) {
    throw new PlatformError(
      `binaries.lock.json entry for '${platformKey}' is incomplete ` +
        "(needs filename, sha256, executable) — likely a placeholder entry with " +
        "no published binary for this platform.",
    );
  }
  return entry;
}

/** True when an entry's sha256 is the not-yet-published placeholder sentinel. */
export function isPlaceholderSha(sha256: string): boolean {
  return /^0+$/.test(sha256) || sha256.toUpperCase().includes("PENDING");
}

File: src/gaia/daemon/sidecars/platform.py
# Copyright(C) 2025-2026 Advanced Micro Devices, Inc. All rights reserved.
# SPDX-License-Identifier: MIT
"""Platform/arch resolution and binary-lock loading (port of platform.ts).

``binaries.lock.json`` is the single source of truth for which artifact to fetch
for the current host and what its SHA-256 must be. Platform keys are
``{os}-{arch}`` normalized to the npm package's key space
(``win32``/``darwin``/``linux`` + ``x64``/``arm64``).

Relocated from ``gaia.ui.email_sidecar.platform`` (issue #2142, T1).
"""

from __future__ import annotations

import json
import platform as _platform
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from gaia.daemon.sidecars.errors import PlatformError

SUPPORTED_PLATFORMS = ("win32-x64", "darwin-arm64", "darwin-x64", "linux-x64")


@dataclass(frozen=True)
class LockEntry:
    filename: str
    sha256: str
    executable: str
    size: Optional[int] = None


@dataclass(frozen=True)
class BinaryLock:
    schema_version: str
    agent_version: str
    base_url: str
    binaries: "dict[str, LockEntry]"


def current_platform_key(plat: Optional[str] = None, arch: Optional[str] = None) -> str:
    """Resolve the host's platform key in the npm package's namespace.

    Maps Python's ``sys.platform``/``platform.machine()`` to the ``{os}-{arch}``
    keys used in binaries.lock.json.
    """
    raw_os = plat if plat is not None else sys.platform
    if raw_os.startswith("win"):
        os_key = "win32"
    elif raw_os == "darwin":
        os_key = "darwin"
    elif raw_os.startswith("linux"):
        os_key = "linux"
    else:
        os_key = raw_os
    raw_arch = (arch if arch is not None else _platform.machine()).lower()
    if raw_arch in ("x86_64", "amd64", "x64"):
        arch_key = "x64"
    elif raw_arch in ("arm64", "aarch64"):
        arch_key = "arm64"
    else:
        arch_key = raw_arch
    return f"{os_key}-{arch_key}"


def default_lock_path() -> Path:
    """Locate the repo's binaries.lock.json (npm package ships the canonical one)."""
    # src/gaia/daemon/sidecars/platform.py -> repo root is parents[4].
    repo_root = Path(__file__).resolve().parents[4]
    return repo_root / "hub" / "agents" / "email" / "npm" / "binaries.lock.json"


def load_lock(lock_path: Optional[Path] = None) -> BinaryLock:
    path = Path(lock_path) if lock_path is not None else default_lock_path()
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as e:
        raise PlatformError(
            f"cannot read the sidecar binary lock at {path}: {e}. This file is "
            "only present in a source checkout (it ships in the agent's npm "
            "package, not the installed Python wheel), so an installed GAIA "
            "resolves the binary from an Agent Hub install instead — reinstalling "
            "the Python wheel will not create this file."
        ) from e
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as e:
        raise PlatformError(
            f"binaries.lock.json at {path} is not valid JSON: {e}"
        ) from e
    binaries = parsed.get("binaries")
    if not isinstance(binaries, dict):
        raise PlatformError(f'binaries.lock.json at {path} is missing a "binaries" map')
    entries = {
        k: LockEntry(
            filename=v.get("filename", ""),
            sha256=v.get("sha256", ""),
            executable=v.get("executable", ""),
            size=v.get("size"),
        )
        for k, v in binaries.items()
    }
    return BinaryLock(
        schema_version=parsed.get("schemaVersion", ""),
        agent_version=parsed.get("agentVersion", ""),
        base_url=parsed.get("baseUrl", ""),
        binaries=entries,
    )


def resolve_entry(lock: BinaryLock, platform_key: str) -> LockEntry:
    entry = lock.binaries.get(platform_key)
    if entry is None:
        available = ", ".join(lock.binaries) or "(none)"
        raise PlatformError(
            f"no sidecar binary for platform '{platform_key}'. Available in "
            f"binaries.lock.json: {available}. Supported targets: "
            + ", ".join(SUPPORTED_PLATFORMS)
        )
    if not entry.sha256 or not entry.filename or not entry.executable:
        raise PlatformError(
            f"binaries.lock.json entry for '{platform_key}' is incomplete "
            "(needs filename, sha256, executable) — likely a placeholder with no "
            "published binary for this platform."
        )
    return entry


def is_placeholder_sha(sha256: str) -> bool:
    """True for the not-yet-published sentinel (all-zeros or a PENDING marker)."""
    s = sha256 or ""
    if not s:
        return False
    return s.strip("0") == "" or "PENDING" in s.upper()
